app/src/main/java/com/stevesoltys/seedvault/transport/backup/PackageService.kt

```diff
@@ -157,9 +157,20 @@ internal fun PackageInfo.isSystemApp(): Boolean {
     return applicationInfo.flags and FLAG_SYSTEM != 0
 }
 
-internal fun PackageInfo.allowsBackup(): Boolean {
+internal fun PackageInfo.allowsBackup(context: Context, userId: Int): Boolean {
     if (packageName == MAGIC_PACKAGE_MANAGER || applicationInfo == null) return false
-    return applicationInfo.flags and FLAG_ALLOW_BACKUP != 0
+
+    // This must be kept in sync with frameworks/base/services/backup/java/com/android/server/backup/utils/BackupEligibilityRules.java
+    val isSystemApp = applicationInfo.flags and FLAG_SYSTEM != 0
+    val ignoreAllowBackup = !isSystemApp
+        && (context.checkSelfPermission(Manifest.permission.LOG_COMPAT_CHANGE)
```

Comment:
Is this needed?

Reply:
The permission check is necessary since it is privileged|signature. Not all deployments would have it granted and calling `CompatChanges.isChangeEnabled(IGNORE_ALLOW_BACKUP_IN_D2D, packageName, UserHandle.of(userId)` without them results in a SecurityException. So we don't ingore allowBackup in those cases